lib/autoproj/ops/configuration.rb

```diff
@@ -314,57 +313,50 @@ def inspect
                 to_s
             end
 
+            # Sort the package sets by dependency order
+            # Package sets that have no dependencies come first,
+            # the local package set (by main configuration) last
             def sort_package_sets_by_import_order(package_sets, root_pkg_set)
-                # The sorting is done in two steps:
-                #  - first, we build a topological order of the package sets
-                #  - then, we insert the auto-imported packages, following this
-                #    topological order, in the user-provided order. Each package is
-                #    considered in turn, and added at the earliest place that fits
-                #    the dependencies
-                topological = Array.new
-                queue = (package_sets.to_a + [root_pkg_set]).uniq
-                until queue.empty?
-                    last_size = queue.size
-                    pending = queue.dup
-                    queue = Array.new
-                    until pending.empty?
-                        pkg_set = pending.shift
-                        if pkg_set.imports.any? { |imported_set| !topological.include?(imported_set) }
-                            queue.push(pkg_set)
-                        else
-                            topological << pkg_set
-                        end
-                    end
-                    if queue.size == last_size
-                        raise ArgumentError, "cannot resolve the dependencies between package sets. There seem to be a cycle amongst #{queue.map(&:name).sort.join(', ')}"
-                    end
+                name_to_pkg_set = Hash.new
+                pkg_sets = Hash.new
+                all_pkgset_deps = Set.new
+
+                package_sets.each do |p|
+                    pkg_sets[p.name] = p.imports.to_a
```

Comment:
My initial remark was related to `pkg_sets` too ... You could (I believe "should") use `p` as the key here.

Reply:
Also, to avoid two loops, maybe `(package_sets.to_a + [root_pkg_set]).each` (arguable)